packages/cosmwasm/src/cw3cosmwasmclient.ts

```diff
@@ -0,0 +1,197 @@
+}
+
+interface StartBeforeNumberPaginationOptions {
+  readonly startBefore?: number;
+  readonly limit?: number;
+}
+
+interface StartAfterNumberPaginationOptions {
+  readonly startAfter?: number;
+  readonly limit?: number;
+}
+
+interface StartAfterStringPaginationOptions {
+  readonly startAfter?: string;
+  readonly limit?: number;
+}
+
+export class Cw3CosmWasmClient extends SigningCosmWasmClient {
+  private readonly cw3ContractAddress: string;
+
+  public constructor(
+    apiUrl: string,
+    senderAddress: string,
+    signer: OfflineSigner,
+    cw3ContractAddress: string,
+    gasPrice?: GasPrice,
+    gasLimits?: Partial<GasLimits<CosmWasmFeeTable>>,
+    broadcastMode?: BroadcastMode,
+  ) {
+    super(apiUrl, senderAddress, signer, gasPrice, gasLimits, broadcastMode);
+    this.cw3ContractAddress = cw3ContractAddress;
+  }
+
+  public getThreshold(): Promise<ThresholdResult> {
+    return this.queryContractSmart(this.cw3ContractAddress, { threshold: {} });
+  }
+
+  public getProposal(proposalId: number): Promise<ProposalResult> {
+    return this.queryContractSmart(this.cw3ContractAddress, { proposal: { proposal_id: proposalId } });
+  }
+
+  public listProposals({ startAfter, limit }: StartAfterNumberPaginationOptions = {}): Promise<
+    ProposalsResult
+  > {
+    return this.queryContractSmart(this.cw3ContractAddress, {
+      list_proposals: {
+        start_after: startAfter,
+        limit: limit,
+      },
+    });
+  }
+
+  public reverseProposals({ startBefore, limit }: StartBeforeNumberPaginationOptions = {}): Promise<
+    ProposalsResult
+  > {
+    return this.queryContractSmart(this.cw3ContractAddress, {
+      reverse_proposals: {
+        start_before: startBefore,
+        limit: limit,
+      },
+    });
+  }
+
+  public getVote(proposalId: number, voter: string): Promise<VoteResult> {
+    return this.queryContractSmart(this.cw3ContractAddress, {
+      vote: {
+        proposal_id: proposalId,
+        voter: voter,
+      },
+    });
+  }
+
+  public listVotes(
+    proposalId: number,
+    { startAfter, limit }: StartAfterStringPaginationOptions = {},
+  ): Promise<VotesResult> {
+    return this.queryContractSmart(this.cw3ContractAddress, {
+      list_votes: {
+        proposal_id: proposalId,
+        start_after: startAfter,
+        limit: limit,
+      },
+    });
+  }
+
+  public getVoter(address: string): Promise<VoterResult> {
+    return this.queryContractSmart(this.cw3ContractAddress, {
+      voter: {
+        address: address,
+      },
+    });
+  }
+
+  public listVoters({ startAfter, limit }: StartAfterStringPaginationOptions = {}): Promise<VotersResult> {
+    return this.queryContractSmart(this.cw3ContractAddress, {
+      list_voters: {
+        start_after: startAfter,
+        limit: limit,
+      },
+    });
+  }
+
+  public createMultisigProposal(
+    title: string,
+    description: string,
+    msgs: ReadonlyArray<Record<string, unknown>>,
```

Comment:
This could be more tightly specified to the set of all CosmosMsg types, but that would exclude custom message types, so fine as this I guess. Maybe nice to add better typing though, and then allow `custom: ReadonlyArray<Record<string, unknown>>`, as we know the shape of `bank`, `staking`, `wasm` queries and it is easy to make typos when creating those.

This should be a simple enhancement